src/PowerShellEditorServices/Session/PromptNest.cs

```diff
@@ -0,0 +1,564 @@
+                return;
+            }
+
+            _frameStack.Push(
+                new PromptNestFrame(
+                    frameType.HasFlag(PromptNestFrameType.Remote)
+                        ? PowerShell.Create()
+                        : PowerShell.Create(RunspaceMode.CurrentRunspace),
+                    NewHandleQueue(),
+                    frameType));
+        }
+
+        /// <summary>
+        /// Dispose of the current <see cref="PromptNestFrame" /> and revert to the previous frame.
+        /// </summary>
+        internal void PopPromptContext()
+        {
+            PromptNestFrame currentFrame;
+            lock (_syncObject)
+            {
+                if (_isDisposed || _frameStack.Count == 1)
+                {
+                    return;
+                }
+
+                _frameStack.TryPop(out currentFrame);
+            }
+
+            currentFrame.Dispose();
+        }
+
+        /// <summary>
+        /// Get the <see cref="PowerShell" /> instance for the current
+        /// <see cref="PromptNestFrame" />.
+        /// </summary>
+        /// <param name="isReadLine">Indicates whether this is for a PSReadLine command.</param>
+        /// <returns>The <see cref="PowerShell" /> instance for the current frame.</returns>
+        internal PowerShell GetPowerShell(bool isReadLine = false)
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            // Typically we want to run PSReadLine on the current nest frame.
+            // The exception is when the current frame is remote, in which
+            // case we need to run it in it's own frame because we can't take
+            // over a remote pipeline through event invocation.
+            if (NestedPromptLevel > 1 && !IsRemote)
+            {
+                return CurrentFrame.PowerShell;
+            }
+
+            return isReadLine ? _readLineFrame.PowerShell : CurrentFrame.PowerShell;
+        }
+
+        /// <summary>
+        /// Get the <see cref="RunspaceHandle" /> for the current <see cref="PromptNestFrame" />.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// The <see cref="CancellationToken" /> that can be used to cancel the request.
+        /// </param>
+        /// <param name="isReadLine">Indicates whether this is for a PSReadLine command.</param>
+        /// <returns>The <see cref="RunspaceHandle" /> for the current frame.</returns>
+        internal RunspaceHandle GetRunspaceHandle(CancellationToken cancellationToken, bool isReadLine)
+        {
+            if (_isDisposed)
+            {
+                return null;
+            }
+
+            // Also grab the main runspace handle if this is for a ReadLine pipeline and the runspace
+            // is in process.
+            if (isReadLine && !_powerShellContext.IsCurrentRunspaceOutOfProcess())
+            {
+                GetRunspaceHandleImpl(cancellationToken, isReadLine: false);
```

Comment:
do you need to return here?

Reply:
No, when PSRL is running in the main runspace it controls the pipeline thread.  So the ReadLine handle *and* the main thread handle need to be obtained.